tdrs-backend/tdpservice/search_indexes/parsers/test/test_data_parsing.py

```diff
@@ -0,0 +1,254 @@
+    """Test trailer preparser with malformed trailers."""
+    status, err = preparser.get_trailer_row(empty_file)
+    assert status is False
+    assert err['preparsing'] == 'File too short or missing trailer.'
+
+    status, err = preparser.get_trailer_row(bad_trailer_file)
+    assert status is False
+    assert err['preparsing'] == 'Trailer length incorrect.'
+
+    with pytest.raises(ValueError) as e_info:
+        logger.debug(preparser.get_trailer_row(bad_trailer_file_2))
+    assert str(e_info.value) == 'Last row is not recognized as a trailer row.'
+
+def spy_count_check(spies, expected_counts):
+    """Run reduce against two lists, returning True if all functions were called the expected number of times."""
+    return reduce(lambda carry, expected: carry == all(expected),
+                  [zip([spy.call_count for spy in spies], expected_counts)],
+                  True)
+    # logger.debug("reduceVal: %s", reduceVal)
+    # for spy, expected in zip(spies, expected_counts):
+    #    logger.debug("%s: spy called %s times\texpected\t%s", spy, spy.call_count, expected)
+    #    assert spy.call_count == expected
+
+@pytest.mark.django_db
+def test_preparser_body(test_file, mocker):
+    """Test that preparse correctly calls lower parser functions...or doesn't."""
+    spy_preparse = mocker.spy(preparser, 'preparse')
+    spy_head = mocker.spy(preparser, 'validate_header')
+    spy_tail = mocker.spy(preparser, 'validate_trailer')
+    spy_parse = mocker.spy(tanf_parser, 'parse')
+    spy_t1 = mocker.spy(tanf_parser, 'active_t1')
+
+    spies = [spy_preparse, spy_head, spy_tail, spy_parse, spy_t1]
+    assert preparser.preparse(test_file, 'TANF', 'Active Case Data')
+
+    assert spy_count_check(spies, [1, 1, 1, 1, 1])
+
+@pytest.mark.django_db
+def test_preparser_big_file(test_big_file, mocker):
+    """Test the preparse correctly handles a large, correct file."""
+    spy_preparse = mocker.spy(preparser, 'preparse')
+    spy_head = mocker.spy(preparser, 'validate_header')
+    spy_tail = mocker.spy(preparser, 'validate_trailer')
+    spy_parse = mocker.spy(tanf_parser, 'parse')
+    spy_t1 = mocker.spy(tanf_parser, 'active_t1')
+
+    spies = [spy_preparse, spy_head, spy_tail, spy_parse, spy_t1]
+    assert preparser.preparse(test_big_file, 'TANF', 'Active Case Data')
+
+    assert spy_count_check(spies, [1, 1, 1, 1, 815])
+
+@pytest.mark.django_db
+def test_preparser_bad_file(bad_test_file, bad_file_missing_header, mocker):
+    """Test that preparse correctly catches issues in a bad file."""
+    spy_preparse = mocker.spy(preparser, 'preparse')
+    spy_head = mocker.spy(preparser, 'validate_header')
+    spy_tail = mocker.spy(preparser, 'validate_trailer')
+    spy_parse = mocker.spy(tanf_parser, 'parse')
+    spy_t1 = mocker.spy(tanf_parser, 'active_t1')
+
+    spies = [spy_preparse, spy_head, spy_tail, spy_parse, spy_t1]
+    is_valid, preparser_errors = preparser.preparse(bad_test_file, 'TANF', 'Active Case Data')
```

Comment:
It looks like this is the only place the preparser return boolean is used. If the return value is not used in the app code it may not be needed.

Reply:
Yes, we'll have to have a discussion around what 1354 needs when validating files and what's useful for saving the ParserLog model. We definitely need a standardized return and I'm not convinced what I've done here is best -- mostly following the lead because that's what Cerberus did (returnCode:bool, errors:dict). It's useful to know if there was a problem by checking the return instead of inspecting the errors because you won't know the sturcture of it apriori.

Reply:
TODO: standardized return function/class, up for architecture discussion with regards to ParserLog